### PR TITLE
[RFC] convert MCHAR etc operator and register types to enum MotionType

### PR DESCRIPTION
**Pull request by bfredl**
part of the clipboard related work ( #4523 etc ) but realized this piece is mostly independent.

### CODE REVIEW

File: src/nvim/eval.c [hunk 1]
```diff
@@ -14456,11 +14456,11 @@ static void f_setreg(typval_T *argvars, typval_T *rettv)
   char_u      *strregname;
   char_u      *stropt;
   bool append = false;
-  char_u yank_type;
+  MotionType yank_type;
   long block_len;
 
   block_len = -1;
-  yank_type = MAUTO;
+  yank_type = kUnknownMotion;
```
Comment (ZyX-I) [inline src/nvim/eval.c]: This should be something like kMotionTypeUnknown, kMotionTypeChar, kMotionTypeBlock, kMotionTypeLine. I.e. with prefix common for all enum members.

Comment (bfredl) [inline src/nvim/eval.c]: `kMotionTypeUnknown` is better yes, but for the others, "wise" is the established form in documentation and elsewhere, and reads in prose like "when register type is kLineWise". We have several new-style enums without common prefix.

Comment (ZyX-I) [inline src/nvim/eval.c]: There is no place for “Unknown” here. Prose is fine with “when register is linewise”, it is not code. Also it is still possible to have common prefix and Wise: kMTUnknown, kMTCharacterWise, …

Comment (ZyX-I) [inline src/nvim/eval.c]: By the way, why did you choose to rename constants? When I converted VAR_LOCKED/… and VAR_NUMBER/… into enums I preferred to keep old names in order not to fight with lots of linter errors like you will have to now.

Comment (bfredl) [inline src/nvim/eval.c]: Dunno,  "when motion_type == kLineWise, ..." in a doc/comment (prose) reads naturally as "when the motion type is linewise" and is at the same time clear with how the condition looks in code. I can add  MT if the style mandates it but I don't see the value (what usage below is unclear) and why then are we already adding enums that don't follow it.

Comment (ZyX-I) [inline src/nvim/eval.c]: It is better to ask people which add enums that do not follow it why. I do not add such enums. Common prefix indicates that value belongs to some enum, it also provides a way to easily search for all places where any enum value is mentioned.

Comment (bfredl) [inline src/nvim/eval.c]: we intend to fix linter errors, no? the new names correspond directly to the terms in the documentation and much discussion.

Comment (bfredl) [inline src/nvim/eval.c]: good point, `kMT' is is then.

Comment (ZyX-I) [inline src/nvim/eval.c]: We do. Though you have better position to do this: my change was part of PR which added JSON support and many unrelated style fixes were not good for it, your PR is dedicated for this change.

File: src/nvim/normal.h [hunk 2]
```diff
@@ -10,18 +10,29 @@
 #define FIND_STRING     2       /* find any string (WORD) */
 #define FIND_EVAL       4       /* include "->", "[]" and "." */
 
+/// Motion types, used for operators and for yank/delete registers.
+///
+/// The three valid numerical values must not be changes, as they
```
Comment (paulburlumi) [inline src/nvim/normal.h]: changes -> changed

Comment (bfredl) [inline src/nvim/normal.h]: :+1:

File: src/nvim/ops.c [hunk 3]
```diff
@@ -5013,24 +5020,24 @@ void write_reg_contents_ex(int name,
...
     y_ptr->y_size = 0;
   }
 
-  int type = yank_type;  // MCHAR, MLINE or MBLOCK
-  if (yank_type == MAUTO) {
-    type = ((str_list ||
-             (len > 0 && (str[len - 1] == NL || str[len - 1] == CAR)))
-            ? MLINE : MCHAR);
+  if (yank_type == kMTUnknown) {
+    yank_type = ((str_list ||
```
Comment (justinmk) [inline src/nvim/ops.c]: Assigning to a parameter is generally discouraged.
